### ISSUE TITLE
Implement discovery via provision watchers

### ISSUE THREAD

**Issue by iain-anderson** [edgexfoundry/device-sdk-c#212]
* Provide a new function to be called by DSs implementing the discovery callback, which performs matching and profile selection and adds the new devices if appropriate
* Implement an internally-scheduled periodic discovery call, with a configuration option to control it
* The DS implementation should be able to trigger discovery itself via a nonblocking function call

**Comment by iain-anderson** [comment]
For each discovered device, the implementation will have to provide to the SDK:
- An autogenerated device name
- The ProtocolProperties of the device
- A description string (may be blank)
- A list of label strings (may be empty)
The SDK will first examine the extent of existing devices. If a device already exists with the same ProtocolProperties as the new device, the new device is discarded.
The SDK will then attempt to match the new device against its list of Provision Watchers. If a match is found, the device will be created, using the parameters given by the implementation, and those specified in the Watcher (ie, the Device Profile and the initial AdminState).

**Comment by iain-anderson** [comment]
The Provision Watcher matching algorithm is as follows.

A Provision Watcher holds a list of "identifiers". These are named properties. In order to match, the candidate device must have ProtocolProperties that match all of these identifiers. The identifier values may be regex-style wildcard strings.

A Provision Watcher also holds a list of "blocking identifiers". The device's ProtocolProperties are compared against these, and if any one of them matches, the device is discarded.

**Comment by iain-anderson** [comment]
Implementation notes

- The mechanism for adding devices filtered by provision watchers should be available to the DS implementation at any time, not just as a response to discovery being requested. This allows for the immediate addition of devices which self-announce, in protocols where such a thing is possible.
- The existing facility to unconditionally add a device should also be retained.

**Comment by tonyespy** [comment]
It would be helpful to turn this into a simple design document akin to Lenny's [registry re-factor design](https://github.com/lenny-intel/edgex-go/blob/RefactorRegistry/adr/Registry%20Refactoring%20Design.md). In its current form the ideas in this issue are spread across four different comments.

It's also hard to talk about implementation and translate this in C as there's already a Go interface defined for Discovery. As such it _might_ be easier to propose this with respect to the Go SDK and then make adjustments if/where required for C? An alternative would be to add language specific sections to the proposal.

**comment1, bullet1**: states that a new function which performs matching and profile selection, and then adds new devices if appropriate should be called by the DS implementation

I don't think the DS implementation should be calling this function, I think it's the other way around, the SDK should use the Discovery interface (if it exists) of a DS impl and call it based on a REST call to the /discovery endpoint or an AutoEvent style trigger. The purpose of this interface is to return a list of "available" devices (or "new" devices) which the SDKs then use to perform the matching and add new devices if needed.

**c1,b2**: this seems inline with discussion so far, although it might require multiple options to control

**c1,b3**: I'm not sure what it this means. The DS impl could at any time trigger discovery by calling one of its own functions? 

**c3**: One or more examples would be extremely helpful here. Also you say "regex-style". Does this mean actual [regex](https://en.wikipedia.org/wiki/Regular_expression) or is it more of a [glob-style matching](https://en.wikipedia.org/wiki/Glob_(programming))?

**Note** - you also should mention the infamous, I deleted the device but it keeps getting re-added scenario. If you delete a device, you must also adjust your provision watcher to ensure it doesn't just get re-added.

Finally a few thoughts about the original design. In Barcelona (and earlier), discovery was always triggered via a device service's **/discovery** endpoint. Typically this was accomplished using Support Scheduler, however it could also be driven by some other client on the system (or external to the system if the device service is accessible externally). The endpoint caused a function provided by the DS implementation to be called. This function's return value was an array of "available" devices, where each device in the array consisted of a protocol-specific map of values that could be used to identify a device. How this function actually implemented discovery of new devices was left as an implementation detail. That said, as designed, this call was synchronous. Since some wireless protocols can take upwards of tens of seconds to even minutes to complete a scan cycle used for device discovery, we should consider re-working our design to handle asynchronous device results.

Perhaps we should separate the trigger from the results by supporting two discovery interface methods:

- trigger-discovery, just returns OK or Error (could optionally provide a callback for discovery-done)
- get-devices - returns a list of all available devices (or just new devices?)

We should also provide a means for the SDK to receive asynchronous notifications of new devices similar to the way asynchronous readings work in device-sdk-go (i.e. a goroutine waits on a channel for readings to be pushed asynchronously). This push could either be a single new device, or it could be a list of new devices which are the result of a previous call by the SDK to the impl's trigger discovery method. Ideally an impl would use the SDK's device cache to only ensure that actual new devices are pushed, so that the SDK doesn't need to do extra work to filter out existing devices.

**Comment by iain-anderson** [comment]
Just to note, there is an existing discovery interface in C also. The intention here is to capture the additional requirements for using provision watchers, so I've omitted considerations such as blocking on the /discovery endpoint (the existing implementation initiates the discovery process and returns immediately)

c1, b1 - disagree, it is less convenient in C to build and tear down lists. For Go, returning a list may well be the right idiomatic thing to do.

c1, b3 - it was suggested that a DS may become aware by protocol-specific means that new devices exist. Saying that, the note on self-announcing devices (c4, b1) could cover this.

c3 - I mean regex. I've updated the comment.

Agree that the documentation needs to cover the notorious scenario.

With regard to filtering existing devices, I'd prefer to do this in the SDK rather than in the implementations.

**Comment by jpwhitemn** [comment]
Per the Device Service Working Group - Iain and Tony were going to off-line the discussion about a document containing this design information.

**Comment by jpwhitemn** [comment]
Also per the Device Service Working Group meeting of 12/9 and for this record and for future work, we are not covering the use case of “movable” devices at this point.  Meaning that in implementing automatic provisioning, we are not yet covering the case where a device is automatically de-provisioned from one device service and provisioned under another device service in nearly simultaneous event.  This would be the case if a wearable sensor, for example, was moving between buildings and the device leaves the scan area of one gateway (and one device service) and enters the scan area of another gateway (and another device service).
Some concerns that need to be addressed (potentially) as part of the use case above that is not being handled today include:
Are there conditions where by the device is left in a state where it is somehow left connected to both device services for some period of time (because it is unsuccessfully de-provisioned on one device service before joining the other)
Are there conditions where by the device is left in a state where it is not connected to any device service (because it is unsuccessfully re-provisioned to the second device service)
How do we allow the deprovision/provision to take place as part of automatic work done in the system but also allow for a simple single transactions to deprovision/provision – what is called reassignment in the document below - to be allowed from a management system (like a 3rd party UI)?
Who is in charge of such a reassignment?  Would it be between device services (device service requesting take over of a device from another device service) or would it all be accomplished by core metadata?
There is also a concern to make sure this type of change/addition does not require another EdgeX major version.
See move potential designs and pros/cons of the approaches at https://wiki.edgexfoundry.org/display/FA/Device+and+Device+SDK+Working+Group?preview=/329488/37913043/Movable_Provisioning_Proposal.pdf.

**Comment by iain-anderson** [comment]
Draft design doc at
https://github.com/iain-anderson/device-sdk-c/blob/designdocs/docs/design/discovery.md

**Comment by iain-anderson** [comment]
Approved design doc at https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/design/legacy-design/device-service/discovery.md

**Comment by iain-anderson** [comment]
Closed via #237 and #239
